File: src/main/java/com/github/sorusclient/client/hud/HUDElement.java
package com.github.sorusclient.client.hud;

import com.github.sorusclient.client.Sorus;
import com.github.sorusclient.client.adapter.MinecraftAdapter;
import com.github.sorusclient.client.setting.Setting;
import com.github.sorusclient.client.setting.SettingConfigurableData;
import com.github.sorusclient.client.setting.SettingContainer;
import com.github.sorusclient.client.setting.Util;
import com.github.sorusclient.client.util.Axis;

import java.util.*;

public abstract class HUDElement implements SettingContainer {

    private final Map<String, Setting<?>> settings = new HashMap<>();

    private final Setting<Map<String, List<AttachType>>> attached;
    private final Setting<Double> x, y;
    private final Setting<Double> offsetX, offsetY;
    private final Setting<Double> scale;

    private final Setting<String> internalId;

    private final String id;

    @SuppressWarnings("unchecked")
    public HUDElement(String id) {
        this.id = id;

        this.register("x", this.x = new Setting<>(0.0));
        this.register("y", this.y = new Setting<>(0.0));
        this.register("offsetX", this.offsetX = new Setting<>(-1.0));
        this.register("offsetY", this.offsetY = new Setting<>(-1.0));
        this.register("scale", this.scale = new Setting<>(1.0));
        this.register("internalId", this.internalId = new Setting<>(""));
        this.register("attached", this.attached = new Setting<>((Class<Map<String, List<AttachType>>>) (Class<?>) Map.class, new HashMap<>()));
    }

    protected void register(String id, Setting<?> setting) {
        this.settings.put(id, setting);
    }

    public void setInternalId(String id) {
        this.internalId.setValue(id);
    }

    public String getInternalId() {
        return internalId.getValue();
    }

    public String getId() {
        return id;
    }

    public void updatePosition(HUDElement attachedElement, double[] screenDimensions) {
        this.updatePosition(attachedElement, screenDimensions, new ArrayList<>());
    }

    private void updatePosition(HUDElement attachedElement, double[] screenDimensions, List<HUDElement> alreadyUpdated) {
        if (!alreadyUpdated.contains(this)) {
            alreadyUpdated.add(this);

            AttachType attachTypeX = null;
            AttachType attachTypeY = null;
            for (AttachType attachType : this.attached.getValue().get(attachedElement == null ? "null" : attachedElement.getInternalId())) {
                if (attachType.getAxis() == Axis.X) {
                    attachTypeX = attachType;
                } else {
                    attachTypeY = attachType;
                }
            }

            double attachedX = attachedElement == null ? screenDimensions[0] / 2 : attachedElement.getX(screenDimensions[0]);
            double attachedWidth = attachedElement == null ? screenDimensions[0] : attachedElement.getScaledWidth();
            double attachedY = attachedElement == null ? screenDimensions[1] / 2 : attachedElement.getY(screenDimensions[1]);
            double attachedHeight = attachedElement == null ? screenDimensions[1] : attachedElement.getScaledHeight();

            double newX = this.getX(screenDimensions[0]);
            if (attachTypeX != null) {
                newX = attachedX + attachTypeX.getOtherSide() * attachedWidth / 2 - attachTypeX.getSelfSide() * this.getScaledWidth() / 2;
            }

            double newY = this.getX(screenDimensions[1]);
            if (attachTypeY != null) {
                newY = attachedY + attachTypeY.getOtherSide() * attachedHeight / 2 - attachTypeY.getSelfSide() * this.getScaledHeight() / 2;
            }

            this.setPosition(
                    newX,
                    newY,
                    screenDimensions);

            for (String hudId : this.attached.getValue().keySet()) {
                HUDElement hud = Sorus.getInstance().get(HUDManager.class).getById(hudId);
                if (hud == null) continue;
                hud.updatePosition(this, screenDimensions, alreadyUpdated);
            }
        }
    }

    public void setScale(double scale) {
        this.scale.setValue(scale);
    }

    public void setPosition(double x, double y, double[] screenDimensions) {
        if (x > screenDimensions[0] * 2 / 3) {
            this.offsetX.setValue(1.0);
        } else if (x > screenDimensions[0] * 1 / 3) {
            this.offsetX.setValue(0.0);
        } else {
            this.offsetX.setValue(-1.0);
        }

        if (y > screenDimensions[1] * 2 / 3) {
            this.offsetY.setValue(1.0);
        } else if (x > screenDimensions[0] * 1 / 3) {
            this.offsetY.setValueRaw(0.0);
        } else {
            this.offsetY.setValue(-1.0);
        }

        this.x.setValue((x + this.getScaledWidth() / 2 * offsetX.getValue()) / screenDimensions[0]);
        this.y.setValue((y + this.getScaledHeight() / 2 * offsetY.getValue()) / screenDimensions[1]);
    }

    public void setPositionSpecific(double x, double y, double offsetX, double offsetY) {
        this.x.setValue(x);
        this.y.setValue(y);
        this.offsetX.setValue(offsetX);
        this.offsetY.setValue(offsetY);
    }

    public double getX(double screenWidth) {
        return -this.offsetX.getValue() * this.getScaledWidth() / 2 + this.x.getValue() * screenWidth;
    }

    public double getY(double screenHeight) {
        return -this.offsetY.getValue() * this.getScaledHeight() / 2 + this.y.getValue() * screenHeight;
    }

    public double getScale() {
        return this.scale.getValue();
    }

    public void addAttached(HUDElement hudElement, AttachType attachType) {
        this.attached.getValue().computeIfAbsent(hudElement == null ? "null" : hudElement.getInternalId(), k -> new ArrayList<>()).add(attachType);
    }

    public void clearStaticAttached(List<HUDElement> alreadyCleared) {
        alreadyCleared.add(this);

        for (String key : new HashMap<>(this.attached.getValue()).keySet()) {
            if (key.equals("null")) {
                this.attached.getValue().remove(key);
            }
        }

        for (String elementId : this.attached.getValue().keySet()) {
            HUDElement hudElement = Sorus.getInstance().get(HUDManager.class).getById(elementId);
            if (!alreadyCleared.contains(hudElement)) {
                hudElement.clearStaticAttached(alreadyCleared);
            }
        }
    }

    public void detach() {
        for (String elementId : this.attached.getValue().keySet()) {
            Sorus.getInstance().get(HUDManager.class).getById(elementId).detachOther(this);
        }

        this.attached.getValue().clear();
    }

    private void detachOther(HUDElement hud) {
        this.attached.getValue().remove(hud.getInternalId());
    }

    public void delete(List<HUDElement> alreadyDeleted) {
        alreadyDeleted.add(this);
        Sorus.getInstance().get(HUDManager.class).remove(this);

        for (String elementId : this.attached.getValue().keySet()) {
            HUDElement element = Sorus.getInstance().get(HUDManager.class).getById(elementId);
            if (!alreadyDeleted.contains(element)) {
                if (element != null) {
                    element.delete(alreadyDeleted);
                }
            }
        }
    }

    public List<String> getAttached() {
        return new ArrayList<>(this.attached.getValue().keySet());
    }

    protected abstract void render(double x, double y, double scale);
    public abstract double getWidth();
    public abstract double getHeight();

    public double getScaledWidth() {
        return this.getWidth() * this.getScale();
    }

    public double getScaledHeight() {
        return this.getHeight() * this.getScale();
    }

    public final void render() {
        double[] screenDimensions = Sorus.getInstance().get(MinecraftAdapter.class).getScreenDimensions();
        this.render(this.getX(screenDimensions[0]) - this.getScaledWidth() / 2, this.getY(screenDimensions[1]) - this.getScaledHeight() / 2, this.getScale());
    }

    public boolean isAttachedTo(HUDElement other) {
        return this.getAttached().contains(other.getInternalId());
    }

    @Override
    public void load(Map<String, Object> settings) {
        for (Map.Entry<String, Object> setting : settings.entrySet()) {
            Setting<?> setting1 = this.settings.get(setting.getKey());
            if (setting1 != null) {
                setting1.setValueRaw(Util.toJava(setting1.getType(), setting.getValue()));
            }
        }
    }

    @Override
    public Map<String, Object> save() {
        Map<String, Object> settingsMap = new HashMap<>();
        for (Map.Entry<String, Setting<?>> setting : this.settings.entrySet()) {
            settingsMap.put(setting.getKey(), Util.toData(setting.getValue().getValue()));
        }
        settingsMap.put("class", this.getClass().getName());
        return settingsMap;
    }

    @Override
    public boolean isShared() {
        return false;
    }

    @Override
    public void setShared(boolean isShared) {

    }

    public void addSettings(List<SettingConfigurableData> settings) {

    }

}

File: src/main/kotlin/com/github/sorusclient/client/hud/HUDElement.kt
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.hud

import com.github.sorusclient.client.adapter.AdapterManager
import com.github.sorusclient.client.setting.Setting
import com.github.sorusclient.client.setting.data.CategoryData
import com.github.sorusclient.client.setting.data.SettingData
import com.github.sorusclient.client.setting.display.DisplayedCategory
import com.github.sorusclient.client.util.Axis

abstract class HUDElement(val id: String) {

    var attached: Setting<MutableMap<String, MutableList<AttachType>>>
    private var x: Setting<Double>
    private var y: Setting<Double>
    private var offsetX: Setting<Double>
    private var offsetY: Setting<Double>
    var scale: Setting<Double>
    var internalId: Setting<String>

    val category = CategoryData()
    val uiCategory = DisplayedCategory(displayName)

    abstract val displayName: String

    init {
        category
            .apply {
                data["x"] = SettingData(Setting(0.0).also { x = it })
                data["y"] = SettingData(Setting(0.0).also { y = it })
                data["offsetX"] = SettingData(Setting(-1.0).also { offsetX = it })
                data["offsetY"] = SettingData(Setting(-1.0).also { offsetY = it })
                data["scale"] = SettingData(Setting(1.0).also { scale = it })
                data["internalId"] = SettingData(Setting("$id-${System.nanoTime() % 1000}").also { internalId = it })
                data["attached"] = SettingData(Setting(MutableMap::class.java as Class<MutableMap<String, MutableList<AttachType>>>, HashMap()).also { attached = it })
            }

        uiCategory.parent = HUDManager.hudDisplayedCategory
    }

    private fun getInternalId(): String {
        return internalId.value
    }

    fun updatePosition(attachedElement: HUDElement?, screenDimensions: DoubleArray) {
        this.updatePosition(attachedElement, screenDimensions, ArrayList())
    }

    private fun updatePosition(
        attachedElement: HUDElement?,
        screenDimensions: DoubleArray,
        alreadyUpdated: MutableList<HUDElement>
    ) {
        if (!alreadyUpdated.contains(this)) {
            alreadyUpdated.add(this)
            var attachTypeX: AttachType? = null
            var attachTypeY: AttachType? = null

            for (attachType in attached.value[attachedElement?.getInternalId() ?: "null"]!!) {
                if (attachType.axis === Axis.X) {
                    attachTypeX = attachType
                } else {
                    attachTypeY = attachType
                }
            }
            val attachedX = attachedElement?.getX(screenDimensions[0]) ?: (screenDimensions[0] / 2)
            val attachedWidth = attachedElement?.scaledWidth ?: screenDimensions[0]
            val attachedY = attachedElement?.getY(screenDimensions[1]) ?: (screenDimensions[1] / 2)
            val attachedHeight = attachedElement?.scaledHeight ?: screenDimensions[1]
            var newX = getX(screenDimensions[0])
            if (attachTypeX != null) {
                newX = attachedX + attachTypeX.otherSide * attachedWidth / 2 - attachTypeX.selfSide * scaledWidth / 2
            }
            var newY = getX(screenDimensions[1])
            if (attachTypeY != null) {
                newY = attachedY + attachTypeY.otherSide * attachedHeight / 2 - attachTypeY.selfSide * scaledHeight / 2
            }
            setPosition(
                newX,
                newY,
                screenDimensions
            )
            for (hudId in this.attached.value.keys) {
                val hud = HUDManager.getById(hudId) ?: continue
                hud.updatePosition(this, screenDimensions, alreadyUpdated)
            }
        }
    }

    fun setPosition(x: Double, y: Double, screenDimensions: DoubleArray) {
        if (x > screenDimensions[0] * 2 / 3) {
            offsetX.realValue = 1.0
        } else if (x > screenDimensions[0] * 1 / 3) {
            offsetX.realValue = 0.0
        } else {
            offsetX.realValue = -1.0
        }
        if (y > screenDimensions[1] * 2 / 3) {
            offsetY.realValue = 1.0
        } else if (x > screenDimensions[0] * 1 / 3) {
            offsetY.setValueRaw(0.0)
        } else {
            offsetY.realValue = -1.0
        }
        this.x.realValue = (x + scaledWidth / 2 * offsetX.value) / screenDimensions[0]
        this.y.realValue = (y + scaledHeight / 2 * offsetY.value) / screenDimensions[1]
    }

    fun getX(screenWidth: Double): Double {
        return -offsetX.value * scaledWidth / 2 + x.value * screenWidth
    }

    fun getY(screenHeight: Double): Double {
        return -offsetY.value * scaledHeight / 2 + y.value * screenHeight
    }

    private fun getScale(): Double {
        return scale.value
    }

    fun addAttached(hudElement: HUDElement?, attachType: AttachType) {
        attached.value.computeIfAbsent(
            hudElement?.getInternalId() ?: "null"
        ) { ArrayList() }.add(attachType)
    }

    fun clearStaticAttached(alreadyCleared: MutableList<HUDElement?>) {
        alreadyCleared.add(this)
        for (key in HashMap(attached.value).keys) {
            if (key == "null") {
                attached.value.remove(key)
            }
        }
        for (elementId in attached.value.keys) {
            val hudElement = HUDManager.getById(elementId)
            if (!alreadyCleared.contains(hudElement)) {
                hudElement!!.clearStaticAttached(alreadyCleared)
            }
        }
    }

    fun detach() {
        for (elementId in attached.value.keys) {
            if (elementId != "null") {
                HUDManager.getById(elementId)!!.detachOther(this)
            }
        }
        attached.value.clear()
    }

    private fun detachOther(hud: HUDElement) {
        attached.value.remove(hud.getInternalId())
    }

    fun delete(alreadyDeleted: MutableList<HUDElement?>) {
        alreadyDeleted.add(this)
        HUDManager.remove(this)
        for (elementId in attached.value.keys) {
            val element = HUDManager.getById(elementId)
            if (!alreadyDeleted.contains(element)) {
                element?.delete(alreadyDeleted)
            }
        }
    }

    private fun getAttached(): List<String> {
        return ArrayList(attached.value.keys)
    }

    protected abstract fun render(x: Double, y: Double, scale: Double)
    abstract val width: Double
    abstract val height: Double
    val scaledWidth: Double
        get() = width * getScale()
    val scaledHeight: Double
        get() = height * getScale()

    fun render() {
        val screenDimensions = AdapterManager.adapter.screenDimensions
        this.render(
            getX(screenDimensions[0]) - scaledWidth / 2, getY(screenDimensions[1]) - scaledHeight / 2,
            getScale()
        )
    }

    fun isAttachedTo(other: HUDElement): Boolean {
        return getAttached().contains(other.getInternalId())
    }

}